### DISCUSSION TITLE
readOnlyRootFilesystem=true

### DISCUSSION THREAD

**Discussion by JSlater743** [kedacore/keda#2880] [Q&A / Need Help]
For security and compliance reasons we are trying to run keda with readOnlyRootFilesystem enabled
FAQs say the following:

![image](https://user-images.githubusercontent.com/72266415/161549156-bec199e2-46f4-4cfe-8473-72bdde68a927.png)

We are struggling to understand why the metrics service needs this certificate? Can anyone please explain, or point to more detailed documentation in this area.

Thanks

**Comment by zroubalik** [comment]
It's for secured communication between the KEDA Metrics Server and Kuberentes apiserver

**Reply by JSlater743** [reply]
thanks @zroubalik, do know of any documentation / detailed instructions on how to manually produce this certificate and deploy it via helm.

I've managed attached a volume to the metric container :
	--set volumes.metricsApiServer.extraVolumes[0].name=keda-volume \
	--set volumes.metricsApiServer.extraVolumeMounts[0].name=keda-volume \
	--set volumes.metricsApiServer.extraVolumeMounts[0].mountPath=/mnt/keda \

Is it possible to tell the Metrics server to make it own certificate on this volume instead of on the root file system, or is it required that I have to manually produce a certificate for it?

Thanks for any help you can give.

**Reply by ygnr** [reply]
Is there a guide or example on how to generate and pass the certificates? @zroubalik 
@JSlater743  Have you made it work?

**Reply by JSlater743** [reply]
hi @ygnr, no we have not found a way to do this. If you find some instructions could you please let me know. Thank you

**Reply by ygnr** [reply]
@JSlater743 Made it work with similar to what you mentioned above. I saw that certificates are created in the folder `/apiserver.local.config/certificates/` so I set the mount path as that folder, like this

--set volumes.metricsApiServer.extraVolumes[0].name=keda-volume
--set volumes.metricsApiServer.extraVolumeMounts[0].name=keda-volume
--set volumes.metricsApiServer.extraVolumeMounts[0].mountPath=/apiserver.local.config/certificates/

and then set the `readOnlyRootFilesystem` to `true` on securityContext of metricServer

**Reply by zroubalik** [reply]
Sorry I have been very busy and missed this, that's correct, you need to mount a volume with certs.

@ygnr would be awesome if you can open a PR to KEDA doc repo with this info, so other users can benefit

**Reply by JSlater743** [reply]
@ygnr Awesome!! that works :-) . Thank you

**Comment by joebowbeer** [comment]
Summarizing:

The FAQ answer says you can’t run KEDA with readOnlyRootFilesystem=true unless you create your own cert and configure metrics server to use it.

However it is possible to run KEDA with readOnlyRootFilesystem=true simply by creating an emptyDir volume (default type) and mounting it to the path where, by default, metrics server writes its generated cert:

`/apiserver.local.config/certificates/`

The helm command is:

```bash
helm install keda kedacore/keda --namespace keda \
  --set 'volumes.metricsApiServer.extraVolumes[0].name=keda-volume' \
  --set 'volumes.metricsApiServer.extraVolumeMounts[0].name=keda-volume' \
  --set 'volumes.metricsApiServer.extraVolumeMounts[0].mountPath=/apiserver.local.config/certificates/' \
  --set 'securityContext.metricServer.readOnlyRootFilesystem=true'
```

PR to improve FAQ answer: https://github.com/kedacore/keda-docs/pull/830

**Comment by joebowbeer** [comment]
Answered?

https://keda.sh/docs/2.15/reference/faq/#how-do-i-run-keda-with-readonlyrootfilesystemtrue
